social_core/backends/id4me.py

```diff
@@ -0,0 +1,270 @@
+"""
+    ID4me OpenID Connect backend, description at: https://id4me.org/for-developers/
+"""
+import datetime
+import json
+import re
+from calendar import timegm
+
+import dns
+import jwt
+import requests
+from dns.resolver import NXDOMAIN, Timeout
+from jose import jwk, jwt
+from jose.jwt import JWTError, JWTClaimsError, ExpiredSignatureError
+from social_core.backends.open_id_connect import OpenIdConnectAuth
+from social_core.exceptions import AuthUnreachableProvider, AuthForbidden, AuthMissingParameter, AuthTokenError
+from social_core.utils import handle_http_errors
+
+
+class ID4meAssociation(object):
+    """ Use Association model to save the client account."""
+
+    def __init__(self, handle, secret='', issued=0, lifetime=0, assoc_type=''):
+        self.handle = handle  # as client_id and client_secret
+        self.secret = secret.encode()  # not use
+        self.issued = issued  # not use
+        self.lifetime = lifetime  # not use
+        self.assoc_type = assoc_type  # as state
+
+    def __str__(self):
+        return self.handle
+
+
+def is_valid_domain(domain):
+    if domain[-1] == ".":
+        domain = domain[:-1]
+    allowed = re.compile("(?!-)[A-Z\d-]{1,63}(?<!-)$", re.IGNORECASE)
+    return all(allowed.match(x) for x in domain.split("."))
+
+
+class ID4meBackend(OpenIdConnectAuth):
+    name = 'id4me'
+    EXTRA_DATA = ['sub', 'iss', 'clp']
+    JWT_DECODE_OPTIONS = dict(verify_at_hash=False)
+
+    def __init__(self, *args, **kwargs):
+        super(ID4meBackend, self).__init__(*args, **kwargs)
+
+    def get_identity_record(self, identity):
+        try:
+            response = dns.resolver.query('_openid.' + identity, 'TXT', lifetime=5).response.answer
+        except NXDOMAIN or Timeout:
+            raise AuthUnreachableProvider(self)
+        if not response:
+            raise AuthUnreachableProvider(self)
+        records = response[0]
+        if not records:
+            raise AuthUnreachableProvider(self)
+        record = records[-1].strings[0].decode()
+        return {item.split("=")[0]: item.split("=")[1] for item in record.split(";")}
+
+    def get_association(self, issuer):
+        try:
+            return self.strategy.storage.association.get(server_url=issuer)[0]
+        except IndexError:
+            return None
+
+    @handle_http_errors
+    def get_key_and_secret(self):
+        iau = self.strategy.session_get(self.name + '_authority')
+        association = self.get_association(iau)
+        if not association:
```

Comment:
there shall be a check, that registration did not expire (client_secret_expires_at)